platform/web/display_server_web.cpp

```diff
@@ -809,6 +818,100 @@ void DisplayServerWeb::_gamepad_callback(int p_index, int p_connected, const Str
 	}
 }
 
+// IME.
+void DisplayServerWeb::ime_callback(int p_type, const char *p_text) {
+	String text = String::utf8(p_text);
+
+#ifdef PROXY_TO_PTHREAD_ENABLED
+	if (!Thread::is_main_thread()) {
+		callable_mp_static(DisplayServerWeb::_ime_callback).bind(p_type, text).call_deferred();
+		return;
+	}
+#endif
+
+	_ime_callback(p_type, text);
+}
+
+void DisplayServerWeb::_ime_callback(int p_type, const String &p_text) {
+	DisplayServerWeb *ds = get_singleton();
+	// Resume audio context after input in case autoplay was denied.
+	OS_Web::get_singleton()->resume_audio();
+
+	switch (p_type) {
+		case 0: {
+			// IME start.
+			ds->ime_text = String();
+			ds->ime_selection = Vector2i();
+			for (int i = ds->key_event_pos - 1; i >= 0; i--) {
+				// Delete last raw keydown event from query.
+				if (ds->key_event_buffer[i].pressed && ds->key_event_buffer[i].raw) {
+					ds->key_event_buffer.remove_at(i);
+					ds->key_event_pos--;
+					break;
+				}
+			}
+			OS::get_singleton()->get_main_loop()->notification(MainLoop::NOTIFICATION_OS_IME_UPDATE);
+			ds->ime_started = true;
+		} break;
+		case 1: {
+			// IME update.
+			if (ds->ime_active && ds->ime_started) {
+				ds->ime_text = p_text;
+				ds->ime_selection = Vector2i(ds->ime_text.length(), ds->ime_text.length());
+				OS::get_singleton()->get_main_loop()->notification(MainLoop::NOTIFICATION_OS_IME_UPDATE);
+			}
+		} break;
+		case 2: {
+			// IME commit.
+			if (ds->ime_active && ds->ime_started) {
+				ds->ime_started = false;
+
+				ds->ime_text = String();
+				ds->ime_selection = Vector2i();
+				OS::get_singleton()->get_main_loop()->notification(MainLoop::NOTIFICATION_OS_IME_UPDATE);
+
+				String text = p_text;
+				for (int i = 0; i < text.length(); i++) {
+					DisplayServerWeb::KeyEvent ke;
+
+					ke.pressed = true;
+					ke.echo = false;
+					ke.raw = false;
+					ke.keycode = Key::NONE;
+					ke.physical_keycode = Key::NONE;
+					ke.key_label = Key::NONE;
+					ke.unicode = text[i];
+					ke.mod = 0;
+
+					if (ds->key_event_pos >= ds->key_event_buffer.size()) {
+						ds->key_event_buffer.resize(1 + ds->key_event_pos);
+					}
+					ds->key_event_buffer.write[ds->key_event_pos++] = ke;
+				}
+			}
+		} break;
+		default:
+			break;
+	}
+}
+
+void DisplayServerWeb::window_set_ime_active(const bool p_active, WindowID p_window) {
+	ime_active = p_active;
+	godot_js_set_ime_active(p_active);
+}
+
+void DisplayServerWeb::window_set_ime_position(const Point2i &p_pos, WindowID p_window) {
+	godot_js_set_ime_position(p_pos.x, p_pos.y);
+}
+
+Point2i DisplayServerWeb::DisplayServerWeb::ime_get_selection() const {
+	return ime_selection;
+}
+
+String DisplayServerWeb::DisplayServerWeb::ime_get_text() const {
+	return ime_text;
+}
```

Comment:
Is double `DisplayServerWeb::DisplayServerWeb::` here intentional?

Reply:
Ah, that seems indeed to be a mistake I didn't spot it during my review.

Reply:
Fixed.